src/rust/ensogl/lib/core/src/gui/component.rs

```diff
@@ -189,31 +189,91 @@ impl<T> display::Object for ShapeView<T> {
 
 
 
+// ==================
+// === Animatable ===
+// ==================
+
+/// Trait that needs to be implemented to animate a struct.
+pub trait Animatable<T>: Debug+Clone+Default+'static {
+    /// State vector. For nice animations a continuous change in state space needs to be reflected
+    /// in a continuous state in the `Animatable` struct. (There should be no discontinuities).
+    type State: inertia::Value;
+    /// Create the entity from the given state vector.
+    fn from_state(state:Self::State) -> T;
+    /// Get the state vector for the given entity.
+    fn to_state(entity:T) -> Self::State;
+}
+
+macro_rules! define_self_animatable {
+    ($type:ty ) => {
+        impl Animatable<$type> for $type {
+            type  State = $type;
+
+            fn from_state(state:Self::State) -> $type {
+                state
+            }
+
+            fn to_state(entity:$type) -> Self::State {
+                entity
+            }
+        }
+    }
+}
+
+define_self_animatable!(f32);
+define_self_animatable!(Vector2);
+define_self_animatable!(Vector3);
+define_self_animatable!(Vector4);
+
+
+
 // =================
 // === Animation ===
 // =================
 
 /// Smart animation handler. Contains of dynamic simulation and frp endpoint. Whenever a new value
 /// is computed, it is emitted via the endpoint.
-#[derive(CloneRef,Derivative,Debug,Shrinkwrap)]
+#[derive(CloneRef,Derivative,Debug)]
 #[derivative(Clone(bound=""))]
 #[allow(missing_docs)]
-pub struct Animation<T> {
-    #[shrinkwrap(main_field)]
-    pub simulator : DynSimulator<T>,
+pub struct Animation<T:Animatable<T>> {
```

Comment:
Oh, this looks bad! `T:Animatable<T>`. You should never have traits which refer to itself this way. You defintely could have written it as `T:Animatable` instead - see my impl for reference